src/libraries/System.Drawing.Common/src/System/Drawing/ImageInfo.cs

```diff
@@ -41,27 +41,37 @@ public ImageInfo(Image image)
                     PropertyItem? frameDelayItem = image.GetPropertyItem(PropertyTagFrameDelay);
 
                     // If the image does not have a frame delay, we just return 0.
-                    //
                     if (frameDelayItem != null)
                     {
                         // Convert the frame delay from byte[] to int
-                        //
                         byte[] values = frameDelayItem.Value!;
-                        Debug.Assert(values.Length == 4 * FrameCount, "PropertyItem has invalid value byte array");
-                        _frameDelay = new int[FrameCount];
-                        for (int i = 0; i < FrameCount; ++i)
+
+                        // On Windows, we get the frame delays for every frame. On Linux, we only get the first frame delay.
+                        // We handle this by treating the frame delays as a repeating sequence, asserting that the sequence
+                        // is fully repeatable to match the frame count.
+                        Debug.Assert(FrameCount % (values.Length / 4) == 0, "PropertyItem has invalid value byte array. The FrameCount should be evenly divisible by a quarter of the byte array's length.");
+
+                        _frameEndTimes = new int[FrameCount];
+
+                        for (int f = 0, i = 0; f < FrameCount; ++f, i += 4)
                         {
-                            _frameDelay[i] = values[i * 4] + 256 * values[i * 4 + 1] + 256 * 256 * values[i * 4 + 2] + 256 * 256 * 256 * values[i * 4 + 3];
+                            if (i == values.Length)
```

Comment:
> if (i == values.Length) [](http://example.com/codeflow?start=27&length=24)

Are we sure we always have values.Length % 4 == 0? I am asking because we are adding 4 every iteration and we are checking with the values.Length equality here. maybe doing i >= values.Length would be safer?

Reply:
Something low-level in gdi+ would be very much broken if we got an array of bytes that isn't divisible by four. But I'm good with asserting that above and then making this code more defensive.